### PR TITLE
Do not save custom images into database documents

### LINKED ISSUE #4136: Do not save custom images' binary data to database
When loading a custom image (both case images and variants images) the image itself gets saved as binary data, and it's a lot of data:

image

<hr>

There is no need to do so, since also the path to the image on disk is saved in the database, and the image could be loaded on the pages on the fly perhaps - I don't think it would slow down a lot -

### PR DESCRIPTION
**Pull request by northwestwitch**
Fix #4136

- Avoid saving custom images (case and variant) into database
- Retrieve custom images to display on case and variant page using their saved path

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

**How to test**:
Run locally using this branch:
- `scout setup demo`
- `scout --demo serve`

**Expected outcome**:
Custom images should be visible both on [case](http://localhost:5000/cust000/643594) and [variant](http://localhost:5000/cust000/643594/412fa9ca391569d19a1613b1ee8107ca) pages.

**Review:**
- [x] code approved by @alkc 
- [x] tests executed by CR

### CODE REVIEW

**Comment by northwestwitch** [comment]
Tested locally using this branch:

image


<hr>

image

**Comment by dnil** [comment]
Nice! Maybe @alkc or @mhkc would like to review
as they are likely to have the highest impact from it?

**Comment by northwestwitch** [comment]
I'll add a test for the new endpoint and I'm ready with this one I think

File: tests/conftest.py [hunk 1]
```diff
@@ -470,7 +470,7 @@ def case_obj(request, parsed_case):
     case["synopsis"] = ""
     case["updated_at"] = parsed_case["analysis_date"]
     case["delivery_report"] = parsed_case["delivery_report"]
-    case["custom_images"] = parsed_case["custom_images"]["case_images"]
```
Comment (northwestwitch) [inline tests/conftest.py:473]: This was an error

**Comment by alkc** [comment]
> Nice! Maybe @alkc or @mhkc would like to review as they are likely to have the highest impact from it?

We can have a look at the PR and see how the changes work with our setup!

**Review by alkc — COMMENTED** [review]
Hej and thank you for your patience! 

I've checked the proposed changes against our setup. There are no issues on our end, save for some required database clean up.

I've confirmed that images show up in both cases and variants in the demo setup.

In addition I've checked that images work for new and existing cases in our setup. For this I only looked at `case_images`.

I added one comment that might be worth addressing under the Code scanning / SonarCloud check failure in cases.views.

File: scout/server/blueprints/cases/views.py [hunk 2]
```diff
@@ -1043,6 +1043,14 @@
                 return redirect("/public/static/file-earmark-x.svg")
 
 
+@cases_bp.route("/<institute_id>/<case_name>/custom_images")
+def host_custom_image_aux(institute_id: str, case_name: str):
+    """Adds absolute path to a custom image path and returns the image."""
+    institute_and_case(store, institute_id, case_name)
+    abs_path: str = os.path.abspath(request.args.get("image_path"))
+    return send_file(abs_path)
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/cases/views.py:1051]: Good point, I'll fix as you suggest!

File: scout/server/blueprints/cases/views.py [hunk 3]
```diff
@@ -1043,6 +1045,26 @@
...
+                yield from custom_images_paths(v)
+            elif isinstance(v, list):
+                for image in v:
+                    yield image.get("path")
+
+    _, case_obj = institute_and_case(store, institute_id, case_name)
+    custom_images_values: list = list(custom_images_paths(case_obj.get("custom_images", {})))
+    if request.args.get("image_path") in custom_images_values:
+        abs_path: str = os.path.abspath(request.args.get("image_path"))
+        return send_file(abs_path)
```
[the remark this answers is not in the document]
Comment (northwestwitch) [reply scout/server/blueprints/cases/views.py:1065]: Code scanning still doesn't like it, but I think we are good after checking that the image path is among the paths saved for images in the case document.

[the remark this answers is not in the document]
Comment (alkc) [reply scout/server/blueprints/cases/views.py:1065]: I think it looks good. And what a nice way to solve it using recursion!

**Review by alkc — APPROVED** [review]
Looks good to me!

The images load as they should in the demo, and I can no longer access any file using the custom_images endpoint.

**Comment by northwestwitch** [comment]
> Looks good to me!
> 
> The images load as they should in the demo, and I can no longer access any file using the custom_images endpoint.

Thanks for your review. I'll merge now. Keep this change in mind when you deploy next Scout release, in case you need to fix anything in your setup (that database cleanup that you mentioned).

**Pull request merged**